File: PO_java/program/World.java
package projekt.program;

import projekt.program.Organisms.Animals.*;
import projekt.program.Organisms.Plants.*;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class World {
    private int height;
    private int width;
    private char[][] fields;
    private List<Organism> organisms;
    private boolean humanAlive;
    private int turn;
    //private String message;

    public World(int height, int width) {
        this.height = height;
        this.width = width;
        this.fields = new char[height][width];
        this.organisms = new ArrayList<>();
        this.humanAlive = true;
        this.turn = 1;
        this.fillBoard();
    }

    private void fillBoard() {
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                fields[y][x] = ' ';
            }
        }
    }

    public void executeTurn(char command) {
        if (!humanAlive) {
            return;
        }
        System.out.println(" ");
        this.turn += 1;
        List<Organism> organismsCopy = new ArrayList<>(organisms);
        for (int j = 7; j >= 0; j--) {
            for (Organism organism : organismsCopy) {
                if (organism.getInitiative() == j && organism.isAlive()) {
                    organism.action(command);
                }
            }
        }
    }

    public int getTurn() {
        return turn;
    }

    public void setTurn(int newTurn){
        this.turn = newTurn;
    }

    public int getHeight() {
        return height;
    }

    public int getWidth(){
        return width;
    }
    public boolean isPositionEmpty(int x, int y) {
        return fields[y][x] == ' ';
    }

    public boolean isPositionValid(int x, int y) {
        return x >= 0 && y >= 0 && x < width && y < height;
    }

    public Organism getOrganismAtPosition(int x, int y) {
        for (Organism organism : organisms) {
            if (organism.getX() == x && organism.getY() == y) {
                return organism;
            }
        }
        return null;
    }

    public void addOrganism(Organism organism) {
        int x, y;
        Random rand = new Random();
        while (true) {
            x = organism.getX();
            y = organism.getY();
            if (isPositionValid(x, y) && isPositionEmpty(x, y)) {
                placeOrganismAtPosition(x, y, organism);
                break;
            } else {
                organism.setPosition(rand.nextInt(width), rand.nextInt(height));
            }
        }
    }

    private void placeOrganismAtPosition(int x, int y, Organism organism) {
        organisms.add(organism);
        fields[y][x] = organism.getSymbol();
    }

    public void changePosition(int x, int y, int px, int py) {
        char tmp = fields[y][x];
        fields[y][x] = ' ';
        fields[y + py][x + px] = tmp;
    }

    public void clearPosition(int x, int y) {
        Organism organism = getOrganismAtPosition(x, y);
        if (organism != null) {
            organisms.remove(organism);
            fields[y][x] = ' ';
        }
    }

    public void killHuman() {
        this.humanAlive = false;
    }

    public void saveToFile() {
        try (PrintWriter writer = new PrintWriter(new File("slot1.txt"))) {
            writer.println(width + " " + height + " " + turn);
            writer.println(organisms.size());
            for (Organism organism : organisms) {
                if(organism instanceof Human){
                    writer.printf("%s %d %d %d %d%n", organism.getName(), organism.getX(), organism.getY(), ((Human) organism).getTurnCounter(), organism.getStrength());
                }
                else writer.printf("%s %d %d %d%n", organism.getName(), organism.getX(), organism.getY(), organism.getStrength());
            }
        } catch (FileNotFoundException e) {
            System.err.println("Error opening file: " + e.getMessage());
        }
    }

    public static World loadFromFile() {
        try (BufferedReader reader = new BufferedReader(new FileReader("slot1.txt"))) {
            String[] firstLine = reader.readLine().split(" ");
            int width = Integer.parseInt(firstLine[0]);
            int height = Integer.parseInt(firstLine[1]);
            int turn = Integer.parseInt(firstLine[2]);

            World world = new World(height, width);
            world.setTurn(turn);

            int numberOfAnimals = Integer.parseInt(reader.readLine());
            for (int i = 0; i < numberOfAnimals; i++) {
                String[] organismData = reader.readLine().split(" ");
                String name = organismData[0];
                int x = Integer.parseInt(organismData[1]);
                int y = Integer.parseInt(organismData[2]);
                int strength = Integer.parseInt(organismData[3]);

                switch (name) {
                    case "Human":
                        int turnCounter = Integer.parseInt(organismData[4]);
                        Human h = new Human(world, x, y);
                        h.setHumanStrenght(strength, turnCounter);
                        break;
                    case "Turtle":
                        Turtle t = new Turtle(world, x, y);
                        t.setOrganismsStrenght(strength);
                        break;
                    case "Fox":
                        Fox fox = new Fox(world, x, y);
                        fox.setOrganismsStrenght(strength);
                        break;
                    case "Sheep":
                        Sheep s = new Sheep(world, x, y);
                        s.setOrganismsStrenght(strength);
                        break;
                    case "Wolf":
                        Wolf wolf = new Wolf(world, x, y);
                        wolf.setOrganismsStrenght(strength);
                        break;
                    case "Antelope":
                        Antelope antelope = new Antelope(world, x, y);
                        antelope.setOrganismsStrenght(strength);
                        break;
                    case "Grass":
                        Grass grass = new Grass(world, x, y);
                        break;
                    case "Dandelion":
                        Dandelion dandelion = new Dandelion(world, x, y);
                        break;
                    case "DeadlyNightshade":
                        DeadlyNightshade deadlyNightshade = new DeadlyNightshade(world, x, y);
                        break;
                    case "Guarana":
                        Guarana guarana = new Guarana(world, x, y);
                        break;
                    case "Barszcz":
                        Barszcz barszcz = new Barszcz(world, x, y);
                        break;
                    default:
                        throw new IllegalArgumentException("Unknown organism type: " + name);
                }
            }
            return world;
        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
            return null;
        }
    }

}

File: PO_py/program/World.py
import random

from program.Organisms.Animals.Antelope import Antelope
from program.Organisms.Animals.CyberSheep import CyberSheep
from program.Organisms.Animals.Fox import Fox
from program.Organisms.Animals.Human import Human
from program.Organisms.Animals.Sheep import Sheep
from program.Organisms.Animals.Turtle import Turtle
from program.Organisms.Animals.Wolf import Wolf
from program.Organisms.Plants.Barszcz import Barszcz
from program.Organisms.Plants.Dandelion import Dandelion
from program.Organisms.Plants.DeadlyNightshade import DeadlyNightshade
from program.Organisms.Plants.Grass import Grass
from program.Organisms.Plants.Guarana import Guarana


class World:
    def __init__(self, height, width):
        self.height = height
        self.width = width
        self.fields = [[' ' for _ in range(width)] for _ in range(height)]
        self.organisms = []
        self.humanAlive = True
        self.turn = 1
        self.barszcz_num = 0;
        self.fill_board()


    def fill_board(self):
        for y in range(self.height):
            for x in range(self.width):
                self.fields[y][x] = ' '

    def execute_turn(self, command):
        if not self.humanAlive:
            return
        print(" ")
        self.turn += 1
        organisms_copy = self.organisms[:]
        for j in range(7, -1, -1):
            for organism in organisms_copy:
                if organism.get_initiative() == j and organism.is_alive():
                    organism.action(command)

    def get_turn(self):
        return self.turn

    def set_turn(self, new_turn):
        self.turn = new_turn

    def get_height(self):
        return self.height

    def get_width(self):
        return self.width

    def is_position_empty(self, x, y):
        return self.fields[y][x] == ' '

    def is_position_valid(self, x, y):
        return 0 <= x < self.width and 0 <= y < self.height

    def add_barszcz(self):
        self.barszcz_num += 1

    def destroy_barszcz(self):
        self.barszcz_num -=1

    def znajdz_najblizszy_barszcz(self, x, y):
        min_distance = float('inf')
        closest_barszcz = None

        for organism in self.organisms:
            if isinstance(organism, Barszcz):
                distance = abs(organism.x - x) + abs(organism.y - y)
                if distance < min_distance:
                    min_distance = distance
                    closest_barszcz = (organism.x, organism.y)

        return closest_barszcz

    def get_organism_at_position(self, x, y):
        for organism in self.organisms:
            if organism.get_x() == x and organism.get_y() == y:
                return organism
        return None

    def add_organism(self, organism):
        while True:
            x = organism.get_x()
            y = organism.get_y()
            if self.is_position_valid(x, y) and self.is_position_empty(x, y):
                self.place_organism_at_position(x, y, organism)
                break
            else:
                organism.set_position(random.randint(0, self.width-1), random.randint(0, self.height-1))

    def place_organism_at_position(self, x, y, organism):
        self.organisms.append(organism)
        self.fields[y][x] = organism.get_symbol()

    def change_position(self, x, y, px, py):
        tmp = self.fields[y][x]
        self.fields[y][x] = ' '
        self.fields[y + py][x + px] = tmp

    def clear_position(self, x, y):
        organism = self.get_organism_at_position(x, y)
        if organism:
            self.organisms.remove(organism)
            self.fields[y][x] = ' '

    def kill_human(self):
        self.humanAlive = False

    def save_to_file(self, filename="slot1.txt"):
        try:
            with open(filename, "w") as file:
                file.write(f"{self.width} {self.height} {self.turn}\n")
                file.write(f"{len(self.organisms)}\n")
                for organism in self.organisms:
                    if isinstance(organism, Human):
                        file.write(f"{organism.get_name()} {organism.get_x()} {organism.get_y()} {organism.get_turn_counter()} {organism.get_strength()}\n")
                    else:
                        file.write(f"{organism.get_name()} {organism.get_x()} {organism.get_y()} {organism.get_strength()}\n")
            print("simulation saved to file 'slot1.txt'")
        except Exception as e:
            print(f"Error saving to file: {e}")

    @staticmethod
    def load_from_file(filename="slot1.txt"):
        try:
            with open(filename, "r") as file:
                lines = file.readlines()
                first_line = lines[0].strip().split(" ")
                width = int(first_line[0])
                height = int(first_line[1])
                turn = int(first_line[2])

                world = World(height, width)
                world.set_turn(turn)

                number_of_animals = int(lines[1].strip())
                for i in range(2, 2 + number_of_animals):
                    organism_data = lines[i].strip().split(" ")
                    name = organism_data[0]
                    x = int(organism_data[1])
                    y = int(organism_data[2])
                    strength = int(organism_data[3])

                    if name == "Human":
                        turn_counter = int(organism_data[4])
                        human = Human(world, x, y)
                        human.set_human_strength(strength, turn_counter)
                    elif name == "Turtle":
                        turtle = Turtle(world, x, y)
                        turtle.set_organism_strength(strength)
                    elif name == "Fox":
                        fox = Fox(world, x, y)
                        fox.set_organism_strength(strength)
                    elif name == "Sheep":
                        sheep = Sheep(world, x, y)
                        sheep.set_organism_strength(strength)
                    elif name == "CyberSheep":
                        sheep = CyberSheep(world, x, y)
                        sheep.set_organism_strength(strength)
                    elif name == "Wolf":
                        wolf = Wolf(world, x, y)
                        wolf.set_organism_strength(strength)
                    elif name == "Antelope":
                        antelope = Antelope(world, x, y)
                        antelope.set_organism_strength(strength)
                    elif name == "Grass":
                        grass = Grass(world, x, y)
                    elif name == "Dandelion":
                        dandelion = Dandelion(world, x, y)
                    elif name == "DeadlyNightshade":
                        deadly_nightshade = DeadlyNightshade(world, x, y)
                    elif name == "Guarana":
                        guarana = Guarana(world, x, y)
                    elif name == "Barszcz":
                        barszcz = Barszcz(world, x, y)
                    else:
                        raise ValueError(f"Unknown organism type: {name}")

                return world
        except Exception as e:
            print(f"Error reading file: {e}")
            return None

    def new_simulation_setup(self):
        Human(self, 1, 1)
        Sheep(self, 1, 1)
        Sheep(self, 1, 1)
        Sheep(self, 1, 1)
        CyberSheep(self, 1, 1)
        Turtle(self, 1, 1)
        Turtle(self, 1, 1)
        Turtle(self, 1, 1)
        Fox(self, 1, 1)
        Fox(self, 1, 1)
        Wolf(self, 1, 1)
        Wolf(self, 1, 1)
        Wolf(self, 1, 1)
        Wolf(self, 1, 1)
        Antelope(self, 1, 1)
        Antelope(self, 1, 1)
        Antelope(self, 1, 1)
        Grass(self, 1, 1)
        Dandelion(self, 1, 1)
        DeadlyNightshade(self, 1, 1)
        Guarana(self, 1, 1)
        Barszcz(self, 1, 1)

    def make_barszcz(self, x, y):
        Barszcz(self, x, y)